### PR TITLE
data-source/aws_db_subnet_group: create aws_db_subnet_group data-source

### LINKED ISSUE #3326: adding support for data lookup of the resource aws_db_subnet_group
### Terraform Version

v0.10.8

### Affected Resource(s)
Please list the resources as a list, for example:
- `aws_db_subnet_group`

### PR DESCRIPTION
**Pull request by maxenglander**
Adds a data source for `aws_db_subnet_group`. Used `aws_db_instance` as a
model for this work. Currently only allows looking up exactly one
database subnet group using `name` as the argument, although the AWS RDS
API also supports using `filters`. Returns all of the attributes listed
on the AWS docs for the `DBSubnetGroup` data type. The `subnets` attribute
contains a list of subnet identifiers.

Resolves #3326

### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" comments, they generate extra noise for pull request followers and do not help prioritize the request

Output from acceptance testing:

```
$ make testacc TEST=./aws TESTARGS="-run=TestAccAWSDbSubnetGroupDataSource_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -parallel 20 -run=TestAccAWSDbSubnetGroupDataSource_basic -timeout 120m
=== RUN   TestAccAWSDbSubnetGroupDataSource_basic
=== PAUSE TestAccAWSDbSubnetGroupDataSource_basic
=== CONT  TestAccAWSDbSubnetGroupDataSource_basic
--- PASS: TestAccAWSDbSubnetGroupDataSource_basic (20.83s)
PASS
ok      github.com/terraform-providers/terraform-provider-aws/aws       20.899s
```

### CODE REVIEW

**Comment by MOURIK** [comment]
Hello,
Any updates regading this request ? thanks in advance.

**Comment by aferg439** [comment]
echoing @MOURIK - any updates?

File: aws/data_source_aws_db_subnet_group.go [hunk 1]
```diff
@@ -0,0 +1,93 @@
+package aws
+
+import (
+	"fmt"
+	"log"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/rds"
+	"github.com/hashicorp/terraform/helper/schema"
```
Comment (anGie44) [inline aws/data_source_aws_db_subnet_group.go:9]:
```suggestion
	"github.com/hashicorp/terraform-plugin-sdk/helper/schema"
```
here we should reference the schema framework as part of terraform-plugin-sdk instead

File: aws/data_source_aws_db_subnet_group.go [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+
+			"description": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"name": {
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
```
Comment (anGie44) [inline aws/data_source_aws_db_subnet_group.go:30]: here we can omit this attribute as the behavior is reserved for managing a [resource](https://www.terraform.io/docs/extend/schemas/schema-behaviors.html#forcenew)

File: aws/data_source_aws_db_subnet_group.go [hunk 3]
```diff
@@ -0,0 +1,93 @@
...
+			},
+
+			"status": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+
+			"subnets": {
```
Comment (anGie44) [inline aws/data_source_aws_db_subnet_group.go:38]: could we use "subnet_ids" here instead to match the schema field in the corresponding resource? will also help in testing as we can make use of `resource.TestCheckResourceAttrPair()`

File: aws/data_source_aws_db_subnet_group.go [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	if len(resp.DBSubnetGroups) < 1 {
+		return fmt.Errorf("Your query returned no results. Please change your search criteria and try again.")
+	}
+	if len(resp.DBSubnetGroups) > 1 {
+		return fmt.Errorf("Your query returned more than one result. Please try a more specific search criteria.")
+	}
+
+	dbSubnetGroup := *resp.DBSubnetGroups[0]
```
Comment (anGie44) [inline aws/data_source_aws_db_subnet_group.go:73]: pointer dereference can be omitted here as we can still safely use the corresponding `*DBSubnetGroup` in the calls below
```suggestion
	dbSubnetGroup := resp.DBSubnetGroups[0]
```

File: aws/data_source_aws_db_subnet_group.go [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+	}
+
+	log.Printf("[DEBUG] Reading DB SubnetGroup: %s", opts)
+
+	resp, err := conn.DescribeDBSubnetGroups(opts)
+	if err != nil {
+		return err
+	}
+
+	if len(resp.DBSubnetGroups) < 1 {
```
Comment (anGie44) [inline aws/data_source_aws_db_subnet_group.go:66]: I believe these 2 length checks can be reduced to a check like in the resource https://github.com/terraform-providers/terraform-provider-aws/blob/b009f7ac157bdea33c902f8070d1cbcb13cf428b/aws/resource_aws_db_subnet_group.go#L122
as the search is by `name` and the aws SDK only allows 1 subnet group to use a particular name (error 400 thrown at Create time https://docs.aws.amazon.com/AmazonRDS/latest/APIReference/API_CreateDBSubnetGroup.html)

File: aws/data_source_aws_db_subnet_group.go [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+	dbSubnetGroup := *resp.DBSubnetGroups[0]
+
+	d.SetId(d.Get("name").(string))
+
+	d.Set("arn", dbSubnetGroup.DBSubnetGroupArn)
+	d.Set("description", dbSubnetGroup.DBSubnetGroupDescription)
+	d.Set("name", dbSubnetGroup.DBSubnetGroupName)
+	d.Set("status", dbSubnetGroup.SubnetGroupStatus)
+
+	var subnets []string
```
Comment (anGie44) [inline aws/data_source_aws_db_subnet_group.go:82]: let's use a slice here similar to
https://github.com/terraform-providers/terraform-provider-aws/blob/376380f46a9349a191e87597fb5c1cb53850f6f3/aws/resource_aws_db_subnet_group.go#L143

File: aws/data_source_aws_db_subnet_group_test.go [hunk 7]
```diff
@@ -0,0 +1,59 @@
+package aws
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform/helper/acctest"
```
Comment (anGie44) [inline aws/data_source_aws_db_subnet_group_test.go:7]: same as above where we should reference `terraform-plugin-sdk`

File: aws/data_source_aws_db_subnet_group_test.go [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+					resource.TestCheckResourceAttrSet("data.aws_db_subnet_group.bar", "vpc_id"),
+				),
+			},
+		},
+	})
+}
+
+func testAccAWSDBSubnetGroupDataSourceConfig(rInt int) string {
+	return fmt.Sprintf(`
+data "aws_availability_zones" "available" {
```
Comment (anGie44) [inline aws/data_source_aws_db_subnet_group_test.go:35]: here we should ensure the subnet group's subnets have the same availability_zone group (an error is thrown in the case that they may be different) so an alternative pattern to use instead is
```
data "aws_availability_zones" "available" {
  state = "available"

  filter {
    name   = "opt-in-status"
    values = ["opt-in-not-required"]
  }
}
```

File: aws/data_source_aws_db_subnet_group_test.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+func TestAccAWSDbSubnetGroupDataSource_basic(t *testing.T) {
+	rInt := acctest.RandInt()
+	resource.ParallelTest(t, resource.TestCase{
+		PreCheck:  func() { testAccPreCheck(t) },
+		Providers: testAccProviders,
+		Steps: []resource.TestStep{
+			{
+				Config: testAccAWSDBSubnetGroupDataSourceConfig(rInt),
+				Check: resource.ComposeAggregateTestCheckFunc(
```
Comment (anGie44) [inline aws/data_source_aws_db_subnet_group_test.go:19]: here we should follow the pattern of using "TestCheckResoureAttrPair() to verify values in the Terraform state for data sources attributes to compare them with their expected resource attributes"..in the cases where the data_source has an attribute different from the resource (e.g. vpc_id and status), using `resource.TestCheckResourceAttrSet` works as is.

File: website/docs/d/db_subnet_group.html.markdown [hunk 10]
```diff
@@ -0,0 +1,35 @@
+---
+layout: "aws"
+page_title: "AWS: aws_db_subnet_group"
+sidebar_current: "docs-aws-datasource-db-subnet-group"
```
Comment (anGie44) [inline website/docs/d/db_subnet_group.html.markdown:4]: i believe this is no longer used and can be omitted...looks like the `awsproviderlinter` may check for these

**Comment by anGie44** [comment]
hi @maxenglander, thank you for submitting this PR! it's off to a great start :) do you mind rebasing as well/resolving conflicts where needed?

**Comment by cmardonespino** [comment]
Im so excited for this feature! What is missing to merge it !?

**Comment by maxenglander** [comment]
@anGie44 thanks for the review and for adding commits to address the issues with the PR. Let me know if there are any other changes I should make.

**Comment by teamterraform** [comment]
## Notification of Recent and Upcoming Changes to Contributions

Thank you for this contribution! There have been a few recent development changes that affect this pull request. We apologize for the inconvenience, especially if there have been long review delays up until now. Please note that this is automated message from an unmonitored account. See the [FAQ](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/FAQ.md) for additional information on the maintainer team and review prioritization.

If you are unable to complete these updates, please leave a comment for the community and maintainers so someone can potentially continue the work. The maintainers will encourage other contributors to use the existing contribution as the base for additional changes as appropriate. Otherwise, contributions that do not receive updated code or comments from the original contributor may be closed in the future so the maintainers can focus on active items.

For the most up to date information about Terraform AWS Provider development, see the [Contributing Guide](https://github.com/terraform-providers/terraform-provider-aws/blob/master/docs/CONTRIBUTING.md). Additional technical debt changes can be tracked with the [`technical-debt` label on issues](https://github.com/terraform-providers/terraform-provider-aws/issues?q=is%3Aopen+is%3Aissue+label%3Atechnical-debt).

As part of updating a pull request with these changes, the most current unit testing and linting will run. These may report issues that were not previously reported.

### Action Required: Terraform 0.12 Syntax

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/8950
Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14417

Version 3 and later of the Terraform AWS Provider, which all existing contributions would potentially be added, only supports Terraform 0.12 and later. Certain syntax elements of Terraform 0.11 and earlier show deprecation warnings during runs with Terraform 0.12. Documentation and test configurations, such as those including deprecated string interpolations (`some_attribute = "${aws_service_thing.example.id}"`) should be updated to the newer syntax (`some_attribute = aws_service_thing.example.id`). Contribution testing will automatically fail on older syntax in the near future. Please see the referenced issues for additional information.

### Action Required: Terraform Plugin SDK Version 2

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14551

The Terraform AWS Provider has been upgraded to the latest version of the Terraform Plugin SDK. Generally, most changes to contributions should only involve updating Go import paths in source code files. Please see the referenced issue for additional information.

### Action Required: Removal of website/aws.erb File

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14712

Any changes to the `website/aws.erb` file are no longer necessary and should be removed from this contribution to prevent merge issues in the near future when the file is removed from the repository. Please see the referenced issue for additional information.

### Upcoming Change of Git Branch Naming

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14292

Development environments will need their upstream Git branch updated from `master` to `main` in the near future. Please see the referenced issue for additional information and scheduling.

### Upcoming Change of GitHub Organization

Reference: https://github.com/terraform-providers/terraform-provider-aws/issues/14715

This repository will be migrating from https://github.com/terraform-providers/terraform-provider-aws to https://github.com/hashicorp/terraform-provider-aws. No practitioner or developer action is anticipated and most GitHub functionality will automatically redirect to the new location. Go import paths including `terraform-providers` can remain for now. Please see the referenced issue for additional information and scheduling.

**Pull request merged**